### DISCUSSION TITLE
CLI: JSON scaffold configuration file

### DISCUSSION THREAD

**Discussion by integral-llc** [linq2db/linq2db#3600] [Q&A]
Where can I see a full sample of ALL options of the JSON configuration file for scaffolding? Right now I have to dig through the source code to figure out things :) 

If not available, how do I define in that Json file a column type mapping?
I need to have
```
class MyEntity
{
  public MyEnum PropName {get; set; }
}
```

need to change the property type from string to MyEnum for a specific table. 
Thx!

**Comment by MaceWindu** [comment]
https://github.com/linq2db/linq2db/discussions/3531#discussioncomment-2854637

to see all options, use `help` command

**Reply by integral-llc** [reply]
@MaceWindu does it mean I cannot use Json configuration file to have the column types mapping? I really don't want to have another project with interceptors to handle :)

**Reply by MaceWindu** [reply]
you don't need a project. just t4 file

**Reply by MaceWindu** [reply]
currently there is no option for such tuning

**Reply by integral-llc** [reply]
how do I combine the Json configuration with the T4 file? can I run them in the same command, or shall I execute one after another? if second, won't the second overwrite the first's output?

**Reply by MaceWindu** [reply]
https://linq2db.github.io/articles/CLI.html#customize-scaffold-with-code

you can pass path to t4 file in colland line or in json
```json
{
	"general": {
		"customize": "<path_to_t4_file>.tt"
	}
}
```

**Reply by integral-llc** [reply]
I did generate the template using 
```
 dotnet linq2db template
```
and included it into the JSON file.
after executing the `dotnet linq2db scaffold -i collectors.db.json` I'm getting this error
```
T4 compilation failed:
        (128,14): error CS1002: ; expected
        (128,19): error CS1519: Invalid token '/' in class, record, struct, or interface member declaration
        (128,19): error CS1519: Invalid token '/' in class, record, struct, or interface member declaration
        (128,35): error CS1002: ; expected
        (224,2): error CS1028: Unexpected preprocessor directive
        (122,108): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' annotations context.
        (122,29): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' annotations context.
        (141,64): error CS0246: The type or namespace name 'EntityModel' could not be found (are you missing a using directive or an assembly reference?)
        (158,73): error CS0246: The type or namespace name 'StoredProcedureModel' could not be found (are you missing a using directive or an assembly reference?)
        (175,71): error CS0246: The type or namespace name 'TableFunctionModel' could not be found (are you missing a using directive or an assembly reference?)
        (190,72): error CS0246: The type or namespace name 'ScalarFunctionModel' could not be found (are you missing a using directive or an assembly reference?)
        (205,75): error CS0246: The type or namespace name 'AggregateFunctionModel' could not be found (are you missing a using directive or an assembly reference?)
        (221,69): error CS0246: The type or namespace name 'AssociationModel' could not be found (are you missing a using directive or an assembly reference?)
        (128,2): error CS0246: The type or namespace name 'region' could not be found (are you missing a using directive or an assembly reference?)
        (128,20): error CS0246: The type or namespace name 'Code' could not be found (are you missing a using directive or an assembly reference?)
        (128,25): warning CS0169: The field 'Interceptors.Generation' is never used
        (128,9): warning CS0169: The field 'Interceptors.Data' is never used

```

Using Rider under MacOS. 
```
dotnet --version        
6.0.202

```

**Reply by MaceWindu** [reply]
Well, if fresh `scaffold.tt` doesn't give you errors - there is something wrong with your edits. Otherwise I would like to see errors with unmodified tt file

**Reply by integral-llc** [reply]
no no, you misunderstood, this happens with the freshly created scaffold.tt. I just generated and that's it, no changes at all.

**Reply by MaceWindu** [reply]
Could you attach tt file? line numbers doesn't make any sense

**Reply by integral-llc** [reply]
[scaffold.tt.zip](https://github.com/linq2db/linq2db/files/8826528/scaffold.tt.zip)
here it is.
```
 dotnet linq2db --version
dotnet linq2db - Linq To DB command-line utilities. Version: 4.0.0.0

```

**Reply by MaceWindu** [reply]
I see. There were issues with generated template in v4. You need to update to 4.0.1 and generate new tt file:
`dotnet tool update -g linq2db.cli`

**Reply by integral-llc** [reply]
all good now, thx!

but please add this feature to the JSON, much easier to work with.
Even with the old version was much easier to change the types of columns.

**Reply by MaceWindu** [reply]
Could you clarify how it was easier? As I know in old templates you still need to write custom code to locate and update column type.

**Reply by integral-llc** [reply]
new 
```
public override void PreprocessEntity(ITypeParser typeParser, EntityModel entityModel)
	{
    if (entityModel.Metadata.Name?.Name == "item_props")
    {
        var dateColumn = entityModel.Columns.Where(c => c.Metadata.Name == "type").Single();
        // create type from name string because CLI tool use .net core 3.1 runtime without DateOnly type
        dateColumn.Property.Type = typeParser.Parse("Collectors.BackEnd.Infrastructure.Enums.ItemPropType", true);
    }
	}
```

old
```
Tables[""].Columns[""].Type = "MyType"; 
```
much shorter in old version.


**desired**
```json
// scaffold.json

...
"dataModel": {
"tables": {
"columns": {
"myColumnName": {
"Type": "MyType"
}
}
}
}
```

this way I don't have to maintain 2 files and no C# code is needed to be written.

**Comment by Metadorius** [comment]
Are there any updates to this? I need to map quite a bit of scaffolded stuff and the new way of doing this seems to be very limiting.

**Reply by MaceWindu** [reply]
Update on what exactly? There will be not many changes except bugfixes for upcoming release in scaffold. We are out of time to implement something substantial as release planned for next week.

**Reply by Metadorius** [reply]
Just asking to know the status, maybe there are some plans to implement it soon or something among the lines. Thanks for the info :)

**Reply by MaceWindu** [reply]
I will try to implement some stuff this weekend for upcoming release, which could be implemented quickly. No promises
